File: src/2023/day5.rs
fn main() {
    // https://adventofcode.com/2023/day/5
    println!("Day 5, 2023: If You Give A Seed A Fertilizer");

    let fn_part2 = match std::env::args().find(|arg| arg == "brute") {
        Some(_) => part2_bruteforce,
        None => part2,
    };

    debug_assert_eq!(35, part1(get_example()));
    debug_assert_eq!(46, fn_part2(get_example()));

    let p1 = part1(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(389056265, p1);

    let p2 = fn_part2(get_input());
    println!("  Part 2: {}", p2);
    assert_eq!(137516820, p2);
}

fn part1(input: &'static str) -> u64 {
    let almanac = Almanac::new(input);
    almanac
        .seeds
        .iter()
        .map(|&s| almanac.get_location(s))
        .min()
        .unwrap()
}

fn part2_bruteforce(input: &'static str) -> u64 {
    let almanac = Almanac::new(input);
    almanac.get_lowest_location_from_seed_ranges_bruteforce()
}

fn part2(input: &'static str) -> u64 {
    let almanac = Almanac::new(input);
    almanac.get_lowest_location_from_seed_ranges()
}

struct Range {
    dest: u64,
    source: u64,
    len: u64,
}

struct Mapping {
    ranges: Vec<Range>,
}

impl Range {
    fn new(line: &str) -> Self {
        let parts: Vec<u64> = line.split(' ').map(|n| n.parse::<u64>().unwrap()).collect();
        Self {
            dest: parts[0],
            source: parts[1],
            len: parts[2],
        }
    }
}

impl Mapping {
    fn new() -> Mapping {
        Mapping { ranges: vec![] }
    }

    fn parse(&mut self, line: &str) {
        self.ranges.push(Range::new(line));
    }

    fn convert(&self, value: u64) -> u64 {
        for range in &self.ranges {
            if value >= range.source {
                let delta = value - range.source;
                if delta < range.len {
                    return range.dest + delta;
                }
            }
        }

        value
    }

    fn unconvert(&self, value: u64) -> u64 {
        for range in &self.ranges {
            if value >= range.dest {
                let delta = value - range.dest;
                if delta < range.len {
                    return range.source + delta;
                }
            }
        }

        value
    }
}

struct Almanac {
    seeds: Vec<u64>,
    mappings: Vec<Mapping>,
}

impl Almanac {
    fn new(input: &str) -> Self {
        let mut almanac = Self {
            seeds: vec![],
            mappings: vec![],
        };

        let parts: Vec<&str> = input.split("\n\n").collect();

        almanac.seeds = parts[0][7..]
            .split(' ')
            .map(|s| s.parse().unwrap())
            .collect();

        for p in parts.iter().skip(1) {
            let mut mapping = Mapping::new();
            for line in p.lines().skip(1) {
                mapping.parse(line);
            }
            almanac.mappings.push(mapping);
        }

        almanac
    }

    fn get_location(&self, seed: u64) -> u64 {
        self.mappings.iter().fold(seed, |value, r| r.convert(value))
    }

    fn get_seed(&self, location: u64) -> u64 {
        self.mappings
            .iter()
            .rev()
            .fold(location, |value, r| r.unconvert(value))
    }

    fn is_valid_seed(&self, seed: u64) -> bool {
        for (start, count) in self.seeds.chunks(2).map(|s| (s[0], s[1])) {
            if seed >= start && seed < start + count {
                return true;
            }
        }

        false
    }

    fn get_lowest_location_from_seed_ranges_bruteforce(&self) -> u64 {
        let mut location: u64 = 0;

        loop {
            if self.is_valid_seed(self.get_seed(location)) {
                return location;
            }

            location += 1;
        }
    }

    fn get_lowest_location_from_seed_ranges(&self) -> u64 {
        for (start, len) in self.seeds.chunks(2).map(|s| (s[0], s[1])) {
            let _v = self.find_min(start, len);
        }
        0
    }

    fn find_min(&self, start: u64, len: u64) -> u64 {
        for m in &self.mappings {
            for r in &m.ranges {
                eprintln!(
                    "s:{}, l:{} -> dst: {}, src:{}, len:{}",
                    start, len, r.dest, r.source, r.len
                );
            }
        }
        0
    }
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2023/day5.txt")
}

fn get_example() -> &'static str {
    "seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4"
}

File: src/2023/day5.cpp
#include "day5.hpp"
#include "../common.hpp"
#include <chrono>
#include <execution>

using namespace std::literals;
using integer = int64_t;

struct Range
{
    integer start;
    integer length;
};

struct MapRange
{
    integer destination;
    integer source;
    integer length;

    MapRange(std::string_view text)
    {
        Tokenizer t(text);
        destination = svtoi<integer>(t.NextToken(' '));
        source = svtoi<integer>(t.NextToken(' '));
        length = svtoi<integer>(t.NextToken(' '));
    }
};

struct Mapping
{
    std::vector<MapRange> ranges;

    Mapping() = default;

    Mapping(std::string_view text)
    {
        for (auto line : Split(text))
        {
            ranges.emplace_back(line);
        }
    }

    [[nodiscard]] integer Convert(integer value) const
    {
        for (auto const &range : ranges)
        {
            if (value >= range.source)
            {
                if (auto const delta = value - range.source; delta < range.length)
                {
                    return range.destination + delta;
                }
            }
        }

        return value;
    }

    [[nodiscard]] integer Unconvert(integer value) const
    {
        for (auto const &range : ranges)
        {
            if (value >= range.destination)
            {
                if (auto const delta = value - range.destination; delta < range.length)
                {
                    return range.source + delta;
                }
            }
        }

        return value;
    }

    [[nodiscard]] std::vector<Range> ConvertRanges(const std::vector<Range> &inputRanges) const
    {
        std::vector<Range> result;

        for (auto input : inputRanges)
        {
            std::vector<Range> unmapped = {input};

            for (auto const &mapRange : ranges)
            {
                std::vector<Range> nextUnmapped;

                for (auto r : unmapped)
                {
                    integer inputStart = r.start;
                    integer inputEnd = r.start + r.length;
                    integer mapStart = mapRange.source;
                    integer mapEnd = mapRange.source + mapRange.length;

                    // No overlap
                    if (inputEnd <= mapStart || inputStart >= mapEnd)
                    {
                        nextUnmapped.push_back(r);
                        continue;
                    }

                    // Left portion before map range
                    if (inputStart < mapStart)
                    {
                        nextUnmapped.push_back({inputStart, mapStart - inputStart});
                        inputStart = mapStart;
                    }

                    // Overlapping portion (gets mapped)
                    integer overlapStart = std::max(inputStart, mapStart);
                    integer overlapEnd = std::min(inputEnd, mapEnd);
                    result.push_back({mapRange.destination + (overlapStart - mapStart), overlapEnd - overlapStart});

                    // Right portion after map range
                    if (inputEnd > mapEnd)
                    {
                        nextUnmapped.push_back({mapEnd, inputEnd - mapEnd});
                    }
                }

                unmapped = std::move(nextUnmapped);
            }

            // Add unmapped portions (pass through unchanged)
            result.insert(result.end(), unmapped.begin(), unmapped.end());
        }

        return result;
    }
};

struct Almanac
{
    std::vector<integer> seeds;

    Mapping seedToSoil;
    Mapping soilToFertilizer;
    Mapping fertilizerToWater;
    Mapping waterToLight;
    Mapping lightToTemperature;
    Mapping temperatureToHumidity;
    Mapping humidityToLocation;

    Almanac(std::string_view text)
    {
        Mapping *mapping = nullptr;

        for (auto line : Split(text))
        {
            if (seeds.empty())
            {
                for (auto seed : Split(line.substr(7), ' '))
                {
                    seeds.push_back(svtoi<integer>(seed));
                }
            }
            else
            {
                if (mapping == nullptr)
                {
                    if (line.empty())
                    {
                        continue;
                    }

                    Tokenizer t(line);
                    auto const name = t.NextToken(' ');

                    if (name == "seed-to-soil")
                    {
                        mapping = &seedToSoil;
                    }
                    else if (name == "soil-to-fertilizer")
                    {
                        mapping = &soilToFertilizer;
                    }
                    else if (name == "fertilizer-to-water")
                    {
                        mapping = &fertilizerToWater;
                    }
                    else if (name == "water-to-light")
                    {
                        mapping = &waterToLight;
                    }
                    else if (name == "light-to-temperature")
                    {
                        mapping = &lightToTemperature;
                    }
                    else if (name == "temperature-to-humidity")
                    {
                        mapping = &temperatureToHumidity;
                    }
                    else if (name == "humidity-to-location")
                    {
                        mapping = &humidityToLocation;
                    }
                }
                else if (line.empty())
                {
                    mapping = nullptr;
                }
                else
                {
                    mapping->ranges.emplace_back(line);
                }
            }
        }
    }

    [[nodiscard]] integer GetLocation(integer seed)
    {
        auto const soil = seedToSoil.Convert(seed);
        auto const fertilizer = soilToFertilizer.Convert(soil);
        auto const water = fertilizerToWater.Convert(fertilizer);
        auto const light = waterToLight.Convert(water);
        auto const temperature = lightToTemperature.Convert(light);
        auto const humidity = temperatureToHumidity.Convert(temperature);
        return humidityToLocation.Convert(humidity);
    }

    [[nodiscard]] integer GetSeed(integer location)
    {
        auto const humidity = humidityToLocation.Unconvert(location);
        auto const temperature = temperatureToHumidity.Unconvert(humidity);
        auto const light = lightToTemperature.Unconvert(temperature);
        auto const water = waterToLight.Unconvert(light);
        auto const fertilizer = fertilizerToWater.Unconvert(water);
        auto const soil = soilToFertilizer.Unconvert(fertilizer);
        return seedToSoil.Unconvert(soil);
    }

    [[nodiscard]] integer GetLowestLocation()
    {
        std::vector<integer> locations(seeds.size());
        std::transform(begin(seeds), end(seeds), begin(locations),
            [this](auto seed)
            {
                return GetLocation(seed);
            });

        return *std::min_element(begin(locations), end(locations));
    }

    [[nodiscard]] bool IsValidSeed(integer seed) const
    {
        auto const lastSeed = end(seeds);
        for (auto iter = begin(seeds); iter != lastSeed; iter += 2)
        {
            auto const start = *iter;
            auto const count = *std::next(iter);

            if (seed >= start && seed < start + count)
            {
                return true;
            }
        }

        return false;
    }

    [[nodiscard]] integer GetLowestLocationFromSeedRanges()
    {
        std::vector<Range> ranges;

        for (size_t i = 0; i < seeds.size(); i += 2)
        {
            ranges.push_back({seeds[i], seeds[i + 1]});
        }

        ranges = seedToSoil.ConvertRanges(ranges);
        ranges = soilToFertilizer.ConvertRanges(ranges);
        ranges = fertilizerToWater.ConvertRanges(ranges);
        ranges = waterToLight.ConvertRanges(ranges);
        ranges = lightToTemperature.ConvertRanges(ranges);
        ranges = temperatureToHumidity.ConvertRanges(ranges);
        ranges = humidityToLocation.ConvertRanges(ranges);

        return std::min_element(ranges.begin(), ranges.end(),
            [](auto const &a, auto const &b)
            {
                return a.start < b.start;
            })
            ->start;
    }
};

static auto GetLowestLocation(std::string_view almanacText)
{
    Almanac almanac(almanacText);
    return almanac.GetLowestLocation();
}

static auto Part1()
{
    return GetLowestLocation(GetInput());
}

static auto GetLowestLocationFromSeedRanges(std::string_view almanacText)
{
    Almanac almanac(almanacText);
    return almanac.GetLowestLocationFromSeedRanges();
}

static auto Part2()
{
    return GetLowestLocationFromSeedRanges(GetInput());
}

int main()
{
    // https://adventofcode.com/2023/day/5
    fmt::println("Day 5, 2023: If You Give A Seed A Fertilizer");

    Assert(35 == GetLowestLocation(example::almanac));
    Assert(46 == GetLowestLocationFromSeedRanges(example::almanac));

    auto const part1 = Part1();
    fmt::print("  Part 1: {}\n", part1);
    Assert(389056265 == part1);

    auto const part2 = Part2();
    fmt::print("  Part 2: {}\n", part2);
    Assert(137516820 == part2);
}
